### PR TITLE
120 UI feature role accessibility

### LINKED ISSUE #120: Do the UI magic features accessible based on the role

### PR DESCRIPTION
**Pull request by chrisrrowland**
Fixes #120 
Feature to Role mapping defined in `feature.ts`
Role name to canvas name mapped in `models.ts`

I haven't yet tested how useGlobals' `isAuthenticated` and `error` would work if there were issues.

### CODE REVIEW

File: ccm_web/client/src/models/models.ts
Comment (ssciolla) [inline ccm_web/client/src/models/models.ts:5]: You might add a comment above or below `RoleEnum` that says what each side is (left is client representation, right is Canvas representation?).

File: ccm_web/client/src/pages/Home.tsx
Comment (ssciolla) [inline ccm_web/client/src/pages/Home.tsx:4]: Combine these imports?

File: ccm_web/client/src/models/models.ts
Comment (ssciolla) [reply ccm_web/client/src/models/models.ts:5]: Any consideration to making the left side not strings? I.e. `AccountAdmin`, so you can always do things like `RoleEnum.AccountAdmin`?
https://www.typescriptlang.org/docs/handbook/enums.html#string-enums

File: ccm_web/client/src/pages/Home.tsx
Comment (ssciolla) [inline ccm_web/client/src/pages/Home.tsx:37]: Though this works, couldn't we also pass the whole `globals` object in as a prop, since it's available in `App`? That saves us an API call.

File: ccm_web/client/src/pages/Home.tsx
Comment (ssciolla) [inline ccm_web/client/src/pages/Home.tsx:51]: I think there's a way we can ensure the header is rendered as a `h1` in the HTML but maintains the `h6` style. I'll check the API, might be the `as` prop.

File: ccm_web/client/src/pages/Home.tsx
Comment (ssciolla) [inline ccm_web/client/src/pages/Home.tsx:75]: I think if we take the approach I suggested above in letting `App` handle fetching globals, some or all of this may not be necessary. You'll be able to assume it's loaded and the user is authenticated if the `Home` component is being rendered.

File: ccm_web/client/src/models/feature.ts
Comment (ssciolla) [inline ccm_web/client/src/models/feature.ts:56]: Split into multiple lines for readability?

File: ccm_web/client/src/models/FeatureUIData.tsx
Comment (ssciolla) [inline ccm_web/client/src/models/FeatureUIData.tsx:87]: So do you think we'll remove these before merging, or do you want to keep them between PRs?

File: ccm_web/client/src/models/FeatureUIData.tsx
Comment (ssciolla) [inline ccm_web/client/src/models/FeatureUIData.tsx:88]: Nit-pick: You could do `const isAuthorized = rolesAllowingAcess.length > 0`, and then reuse the variable here and in the logging check. This doesn't change anything, but might make the code a little tidier.

File: ccm_web/client/src/models/FeatureUIData.tsx
Comment (ssciolla) [inline ccm_web/client/src/models/FeatureUIData.tsx:80]: I might change this function name to `hasAuthorizedRole`, or something similar. `isAuthorizedForRoles` sounds strange to me, is hard to grok.

File: ccm_web/client/src/pages/Home.tsx
Comment (ssciolla) [inline ccm_web/client/src/pages/Home.tsx:112]: Style thing, I'd unindent these lines. You could also indent `{renderCourseRename}`, but the empty `<>`s get stripped out after (though are still necessary), so putting them all on the same line kinda is a reminder of that.

**Review by ssciolla — COMMENTED** [review]
Looking good, @chrisrrowland! I'm going to test next, but I'm releasing these comments first. I'll approve if I don't see any issues running locally. I do think passing `globals` from `App` will help streamline some code in `Home`. Let me know if you have questions about that.

File: ccm_web/client/src/pages/Home.tsx
Comment (ssciolla) [reply ccm_web/client/src/pages/Home.tsx:51]: It's `component`, see https://material-ui.com/api/typography/#props. I think we want these section headers to be `h2` but have `h6` styling and the course name to be `h1` and have `h5` styling.

**Comment by ssciolla** [comment]
Okay, I tested locally with myself as a user and masquerading as an instructor. Role-based feature rendering seems to be working! I did not test the other roles, but can if someone wants me to.

File: ccm_web/client/src/pages/Home.tsx
Comment (ssciolla) [reply ccm_web/client/src/pages/Home.tsx:37]: @chrisrrowland, I'd like to at minimum discuss this before I approve. If you convince me it's okay as is, I'll approve then.

Comment (chrisrrowland) [reply ccm_web/client/src/pages/Home.tsx:37]: I thought there was some magic to useGlobals that was caching information. I can change the how the code works to avoid more instances of useGlobals

File: ccm_web/client/src/models/models.ts
Comment (chrisrrowland) [reply ccm_web/client/src/models/models.ts:5]: The left side was intended to mirror how they are displayed in Canvas, I can mention that in a comment.

File: ccm_web/client/src/models/FeatureUIData.tsx
Comment (chrisrrowland) [reply ccm_web/client/src/models/FeatureUIData.tsx:87]: I think it will be useful for anyone doing testing to be able to look at the console to see why they were able or not able to see a feature.  I suppose I'd just leave it and have a cleanup issue to remove it later?

File: ccm_web/client/src/models/FeatureUIData.tsx
Comment (chrisrrowland) [reply ccm_web/client/src/models/FeatureUIData.tsx:80]: The naming is kind of consistent  with the other helper functions.  Do you expect them all to be renamed?

File: ccm_web/client/src/pages/Home.tsx
Comment (chrisrrowland) [reply ccm_web/client/src/pages/Home.tsx]: @ssciolla ok I passed globals in to the props for home and removed this

File: ccm_web/client/src/models/FeatureUIData.tsx
Comment (ssciolla) [reply ccm_web/client/src/models/FeatureUIData.tsx:87]: Sure, that's fine with me.

File: ccm_web/client/src/models/models.ts
Comment (ssciolla) [reply ccm_web/client/src/models/models.ts:8]: Ah, didn't realize that. Comment would be helpful, thanks.

File: ccm_web/client/src/models/FeatureUIData.tsx
Comment (chrisrrowland) [reply ccm_web/client/src/models/FeatureUIData.tsx:87]: Created #134

File: ccm_web/client/src/models/FeatureUIData.tsx
Comment (ssciolla) [reply ccm_web/client/src/models/FeatureUIData.tsx:80]: No, I like the other names, what they are doing is clear to me. This one just isn't as clear to me on a language level. How can one be authorized for roles? The function is checking whether a user has one or more roles that authorize them to access a feature. So `hasAuthorizedRole` or `hasAuthorizedRoles` seems to better encapsulate the logic, in my opinion.

I won't insist on this change, I just thought I'd share my opinion.

File: ccm_web/client/src/pages/Home.tsx
Comment (ssciolla) [reply ccm_web/client/src/pages/Home.tsx]: Cool, I will re-review and run it again locally, thanks

File: ccm_web/client/src/models/FeatureUIData.tsx
Comment (chrisrrowland) [reply ccm_web/client/src/models/FeatureUIData.tsx:80]: Ok I can see how that's confusing. I added this function to support course rename since it's the only feature that doesn't fit the pattern so it's kind of a snowflake.  Does `isAuthorizedForRole` work better?

Comment (ssciolla) [reply ccm_web/client/src/models/FeatureUIData.tsx:80]: Yeah, I realized that's why the function was necessary when I was thinking about suggesting combining it with `isAuthorizedForFeature`. I wonder if there's a way we could make course renaming a feature without giving it all the same properties as the others. Might be more trouble than it's worth.

To be completely honest, I don't think `isAuthorizedForRole` is much of an improvement, but I dunno, maybe I'm being unreasonable 😆

File: ccm_web/client/src/App.tsx
Comment (ssciolla) [reply ccm_web/client/src/App.tsx:88]: Looks like you can also remove the `Globals` import from `App` if you do this.

File: ccm_web/client/src/models/FeatureUIData.tsx
Comment (ssciolla) [inline ccm_web/client/src/models/FeatureUIData.tsx:81]: I guess it might make more sense to use `userRole` for parameter in the arrow function?

**Review by ssciolla — APPROVED** [review]
Cool, thanks for making these changes! I couple more comments popped in my head, do what you will with them.

File: ccm_web/client/src/pages/Home.tsx
Comment (ssciolla) [reply ccm_web/client/src/pages/Home.tsx]: Not sure what magic you mean. `useGlobals` just makes the call once, compartmentalizes some of the globals fetching, processing, and authentication check logic.

File: ccm_web/client/src/App.tsx
Comment (chrisrrowland) [reply ccm_web/client/src/App.tsx:88]: I feel like I must be missing some trivial way to do branching inside of JMX but the only way I k now how is to break it out into function calls every time.

File: ccm_web/client/src/models/FeatureUIData.tsx
Comment (chrisrrowland) [reply ccm_web/client/src/models/FeatureUIData.tsx:81]: Thanks for catching that, I renamed the parameters and missed that.

**Pull request merged**
